builtin/credential/aws-ec2/client.go

```diff
@@ -97,28 +118,41 @@ func (b *backend) clientEC2(s logical.Storage, region string) (*ec2.EC2, error)
 	defer b.configMutex.Unlock()
 
 	// If the client gets created while switching the locks, return it
-	if b.EC2ClientsMap[region] != nil {
-		return b.EC2ClientsMap[region], nil
+	if b.EC2ClientsMap[region] != nil && b.EC2ClientsMap[region][stsRole] != nil {
+		return b.EC2ClientsMap[region][stsRole], nil
 	}
 
 	// Create an AWS config object using a chain of providers
-	awsConfig, err := b.getClientConfig(s, region)
+	var awsConfig *aws.Config
+	var err error
+	// The empty stsRole signifies the master account
+	if stsRole == "" {
+		awsConfig, err = b.getClientConfig(s, region)
+	} else {
+		awsConfig, err = b.getStsClientConfig(s, region, stsRole)
+	}
+
 	if err != nil {
 		return nil, err
 	}
 
 	// Create a new EC2 client object, cache it and return the same
-	b.EC2ClientsMap[region] = ec2.New(session.New(awsConfig))
-	return b.EC2ClientsMap[region], nil
+	if _, ok := b.EC2ClientsMap[region]; !ok {
```

Comment:
Unrelated to your changes per se, but I think we should create the client (`ec2.New()`) before this block and do a `nil` check on that.